Lib/test/test_pathlib.py

```diff
@@ -166,6 +166,31 @@ def test_constructor_common(self):
         self.assertEqual(P(P('a'), P('b')), P('a/b'))
         self.assertEqual(P(P('a'), P('b'), P('c')), P(FakePath("a/b/c")))
 
+    def test_bytes(self):
+        P = self.cls
+        with self.assertRaises(TypeError):
+            P(b'a')
+        with self.assertRaises(TypeError):
+            P(b'a', 'b')
+        with self.assertRaises(TypeError):
+            P('a', b'b')
+        with self.assertRaises(TypeError):
+            P('a').joinpath(b'b')
+        with self.assertRaises(TypeError):
+            P('a') / b'b'
+        with self.assertRaises(TypeError):
+            b'a' / P('b')
+        with self.assertRaises(TypeError):
+            P('a').match(b'b')
+        with self.assertRaises(TypeError):
+            P('a').relative_to(b'b')
+        with self.assertRaises(TypeError):
+            P('a').with_name(b'b')
+        with self.assertRaises(TypeError):
+            P('a').with_stem(b'b')
+        with self.assertRaises(TypeError):
+            P('a').with_suffix(b'b')
```

Comment:
Maybe uses `assertRaisesRegex` for these?

Reply:
Mixing strings and bytes currently raises a TypeError with a somewhat misleading message: "Can't mix strings and bytes". In fact, pathlib doesn't support bytes at all, though it may in future.

Should I write a test that checks for that message? Or should I catch the TypeError from the call to `os.path.join()` and re-raise it with a different message? Or just leave this as a little imperfect for now?

Reply:
If the message is imperfect, then we shouldn't assert incorrect behaviour, so for those ones, let's stick with `assertRaises`. For the ones where we have error messages we're happy with, let's use `assertRaisesRegex`.

(And yes, let's open a new issue to track improving the error message here; we can surely do better! But let's not get too distracted by that now ;)

Reply:
OK, I've added an `assertRaisesRegex()` test case for passing in bytes. I'll log a bug about the TypeError message once this PR lands! Thanks for your help :)